ramips: add support for Amped Wireless ALLY router and extender

jonathan-sturges opened the pull request:
Amped Wireless ALLY is a whole-home WiFi kit,
with a router (model ALLY-R1900K) and an
Extender (model ALLY-00X19K).
Both are devices are 11ac and based on
MediaTek MT7621AT and MT7615N chips.  The units
are nearly identical, except the Extender lacks
a USB port and has a single Ethernet port.

Specification:
- SoC: MediaTek MT7621AT (2C/4T) @ 880MHz
- RAM: 128MB DDR3 (Nanya NT5CC64M16GP-DI)
- FLASH: 128MB NAND (Winbond W29N01GVSIAA)
- WiFi: 2.4/5 GHz 4T4R
  - 2.4GHz MediaTek MT7615N bgn
  - 5GHz MediaTek MT7615N nac
- Switch: SoC integrated Gigabit Switch
- USB: 1x USB3 (Router only)
- BTN: Reset, WPS
- LED: single RGB
- UART:  through-hole on PCB.
   J1: pin1 (square pad, towards rear)=3.3V, pin2=RX,
   pin3=GND, pin4=TX.  Settings: 57600/8N1.

Note regarding dual system partitions
-------------------------------------

The vendor firmware and boot loader use a dual partition
scheme.  The boot partition is decided by the bootImage
U-boot environment variable: 0 for the 1st partition, 1
for the 2nd.

OpenWrt does not support this scheme and will always use the
first OS partition.  It will set bootImage to 0 during
installation, making sure the first partition is selected
by the boot loader.

Also, because we can't be sure which partition is active to
begin with, a 2-step flash process is used.  We first
flash an initramfs image, then follow with a
regular sysupgrade.

Installation:

Router (ALLY-R1900K)
1) Install the flashable initramfs image via the OEM web-interface.
  You can use WiFi or Ethernet.
  The direct URL is:  http://192.168.3.1/07_06_00_firmware.html
  a. No login is needed, and you'll be in their setup wizard.
  b. You might get a warning about not being connected to the Internet.
  c. Towards the bottom of the page will be a section entitled "Or
  Manually Upgrade Firmware from a File:"
  where you can manually choose and upload a firmware file.
  d: Click "Choose File", select the OpenWRT "initramfs" image
  and click "Upload."
2) The Router will flash the OpenWrt initramfs image and reboot.  After
  booting, LuCI will be available on 192.168.1.1.
3) Log into LuCI as root; there is no password.
4) Optional (but recommended) is to backup the OEM firmware before
  continuing; see process below.
5) Complete the Installation by flashing a full OpenWRT image.  Note:
  you may use the sysupgrade command line tool in lieu of the UI if
  you prefer.
  a.  Choose System -> Backup/Flash Firmware.
  b.  Click "Flash Image..." under "Flash new firmware image"
  c.  Click "Browse..." and then select the sysupgrade file.
  d.  Click Upload to upload the sysupgrade file.
  e.  Important:  uncheck "Keep settings and retain the current
      configuration" for this initial installation.
  f.  Click "Continue" to flash the firmware.
  g.  The device will reboot and OpenWRT is installed.

Extender (ALLY-00X19K)
1) This device requires a TFTP recovery procedure to do an initial load
  of OpenWRT.  Start by configuring a computer as a TFTP client:
  a. Install a TFTP client (server not necessary)
  b. Configure an Ethernet interface to 192.168.1.x/24; don't use .1 or .6
  c. Connect the Ethernet to the sole Ethernet port on the X19K.
2) Put the ALLY Extender in TFTP recovery mode.
  a. Do this by pressing and holding the reset button on the bottom
  while connecting the power.
  b. As soon as the LED lights up green (roughly 2-3 seconds), release
  the button.
3) Start the TFTP transfer of the Initramfs image from your setup machine.
For example, from Linux:
tftp -v -m binary 192.168.1.6 69 -c put initramfs.bin
4) The Extender will flash the OpenWrt initramfs image and reboot.  After
booting, LuCI will be available on 192.168.1.1.
5) Log into LuCI as root; there is no password.
6) Optional (but recommended) is to backup the OEM firmware before
  continuing; see process below.
7) Complete the Installation by flashing a full OpenWRT image.  Note:
  you may use the sysupgrade command line tool in lieu of the UI if
  you prefer.
  a.  Choose System -> Backup/Flash Firmware.
  b.  Click "Flash Image..." under "Flash new firmware image"
  c.  Click "Browse..." and then select the sysupgrade file.
  d.  Click Upload to upload the sysupgrade file.
  e.  Important:  uncheck "Keep settings and retain the current
      configuration" for this initial installation.
  f.  Click "Continue" to flash the firmware.
  g.  The device will reboot and OpenWRT is installed.

Backup the OEM Firmware:
-----------------------

There isn't any firmware released for the ALLY devices on
the Amped Wireless web site. Reverting back to the OEM firmware is
not possible unless we have a backup of the original OEM
firmware.

The OEM firmware may be stored on either /dev/mtd3 ("firmware")
or /dev/mtd6 ("oem").  We can't be sure which was overwritten
with the initramfs image, so backup both partitions to be safe.

  1) Once logged into LuCI, navigate to System -> Backup/Flash Firmware.
  2) Under "Save mtdblock contents," first select "firmware" and click
  "Save mtdblock" to download the image.
  3) Repeat the process, but select "oem" from the pull-down menu.

Revert to the OEM Firmware:
--------------------------
* U-boot TFTP:
  Follow the TFTP recovery steps for the Extender, and use the
  backup image.

* OpenWrt "Flash Firmware" interface:
  Upload the backup image and select "Force update"
  before continuing.

Signed-off-by: Jonathan Sturges <jsturges@redhat.com>

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally-00x19k.dts:
```diff
@@ -0,0 +1,21 @@
+
+&switch0 {
+	ports {
+
```
drop empty line before and after port@2 here.

jonathan-sturges replied:
ok

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally-r1900k.dts:
```diff
@@ -0,0 +1,36 @@
+	};
+};
+
+
```
drop empty lines at EOF

jonathan-sturges replied:
ok

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally.dtsi:
```diff
@@ -0,0 +1,201 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
```
dts-v1 should only be in DTS, not DTSI.

jonathan-sturges replied:
ok thanks

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally.dtsi:
```diff
@@ -0,0 +1,201 @@
+
+	chosen {
+		bootargs = "console=ttyS0,57600";
+	};
```
chosen can be dropped, it's already in mt7621.dtsi

jonathan-sturges replied:
ok

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally.dtsi:
```diff
@@ -0,0 +1,201 @@
+		};
+
+		led {
+			label = "led";
```
"led" what?

jonathan-sturges replied:
it's just a push-on/push-off switch to toggle the lone LED on/off in the OEM firmware.  I renamed it "led_switch" for clarity.

adschm replied:
If it's a switch you need `linux,input-type = <EV_SW>;` in addition.

jonathan-sturges replied:
thanks, missed that.  think it's correct now.

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally.dtsi:
```diff
@@ -0,0 +1,201 @@
+			label = "ally:red:status";
+			gpios = <&gpio 12 GPIO_ACTIVE_HIGH>;
+		};
+
```
drop empty line here.

jonathan-sturges replied:
ok

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally.dtsi:
```diff
@@ -0,0 +1,201 @@
+			gpios = <&gpio 14 GPIO_ACTIVE_HIGH>;
+		};
+
+		led_status_red: status_red {
```
Please sort leds and keys by something, either gpio number or alphabetic nodes.

jonathan-sturges replied:
sorted by GPIO

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally.dtsi:
```diff
@@ -0,0 +1,201 @@
+
+&state_default {
+	gpio {
+		groups = "uart2", "uart3", "i2c";
```
Do you need i2c? And jtag/wdt seem missing ...

jonathan-sturges replied:
thanks, I removed i2c and added jtag and wdt.

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally.dtsi:
```diff
@@ -0,0 +1,201 @@
+};
+
+&xhci {
+	status = "enabled";
```
1. enabled is no valid keyword
2. this is enabled by default
3. it should only be enabled for "router"

so, you need

```
&xhci {
	status = "disabled";
}
```
for the device without USB _only_.

jonathan-sturges replied:
got it, thx

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally.dtsi:
```diff
@@ -0,0 +1,201 @@
+};
+
+&gmac0 {
+	mtd-mac-address = <&factory 0x4>;
```
Nothing in 0x28, 0x2e, 0xe000, 0xe006?

jonathan-sturges replied:
only the radio MACs are in 'factory' in this unit; all MACs are in 'hwconfig', but that partition is ASCII and easier to grab in 02_network.

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally.dtsi:
```diff
@@ -0,0 +1,201 @@
+
+			partition@400000 {
+				label = "ubi";
+				reg = <0x400000 0x1C00000>;
```
I prefer lower-case hex.

jonathan-sturges replied:
ok

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally.dtsi:
```diff
@@ -0,0 +1,201 @@
+		};
+	};
+};
+
```
drop empty line at eof

jonathan-sturges replied:
ok

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally.dtsi:
```diff
@@ -0,0 +1,201 @@
+				label = "ubi";
+				reg = <0x400000 0x1C00000>;
+			};
+
```
drop empty line here.

jonathan-sturges replied:
ok

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally.dtsi:
```diff
@@ -0,0 +1,201 @@
+				reg = <0x400000 0x1C00000>;
+			};
+
+		};
```
Will this nested scheme with "firmware" and "ubi" keywords work properly?

jonathan-sturges replied:
It works fine as I'm not using auto-splitting.

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -140,6 +140,38 @@ define Device/alfa-network_quad-e4g
 endef
 TARGET_DEVICES += alfa-network_quad-e4g
 
+define Device/ampedwireless_ally-r1900k
```
Please create a common definition Device/ampedwireless_ally to inherit the others from

jonathan-sturges replied:
good idea, I missed that... thanks.

adschm commented on target/linux/ramips/mt7621/base-files/etc/board.d/02_network:
```diff
@@ -30,6 +31,7 @@ ramips_setup_interfaces()
 	ubnt,unifi-nanohd)
 		ucidef_set_interface_lan "lan"
 		;;
+	ampedwireless,ally-r1900k|\
```
Move blocks due to alphabetic sorting.

jonathan-sturges replied:
I'm not following you on this.  Within each port configuration, the device names are already sorted alphabetically.

adschm replied:
and all the blocks as a whole are sorted based on their first entry ...

jonathan-sturges replied:
ahh ok!

adschm commented on target/linux/ramips/mt7621/base-files/etc/board.d/02_network:
```diff
@@ -75,6 +77,15 @@ ramips_setup_macs()
 
 	case $board in
+	ampedwireless,ally-00x19k)
+		lan_mac=$(mtd_get_mac_ascii hwconfig HW.LAN.MAC.Address)
```
Either you set the correct address for gmac in DTS or use lan_mac here. DTS is preferred.
But please don't do both.

Same goes for label-mac-device vs. label_mac here.

jonathan-sturges replied:
ok thanks

vj4git commented:
Is it possible to add the router model ALLY 1200 (model AR1200L) with this merge. The hardware is similar to the above R-1900K, except that it doesn't have the usb. See https://www.ampedwireless.com/media/docs/usersguide/ALLY12B_ALLY19B_UsersGuide.pdf). 

So we need one change in router dtsi
------------
&xhci {
	status = "disabled";
}
------------

Associated extender for ALLY1200 is  identical to the  ALLY-00X19K, so no further changes needed.

jonathan-sturges commented:
I just did a quick review of the AR1200L, and it does indeed seem to be very similar to the R1900K.  However, the devil is in the details, as they say.  Two areas of concern would the NAND layout and boot loader.  The safest bet is still to solder header pins for the console of a AR1200L and capture a boot log of the OEM firmware, for starters.  If it is as similar to the R1900K as it looks to be, then it would be trivial to add support for it after this PR is merged.

vj4git commented:
My research was based on their FCC filings, specifically https://fccid.io/ZTT-AR1200L/Internal-Photos/Internal-Photos-3512404. NAND layout seems to be matching with the https://fccid.io/ZTT-AR1900L. 

Sure bootlog details would be ideal.

jonathan-sturges commented:
Yeah I was reviewing the FCC filings as well.
Don't get me wrong, I think it's likely they're nearly the same devices, but I'd hate to be wrong and soft-brick someone's router.  You could clone my working branch, build for the R1900K target and flash just the initramfs image.  That should tell you if they're as close as they seem to be, and it should leave one of the OEM images untouched.

vj4git commented:
Yes, that's exactly what I was planning to do - clone yours and build for the AR1200L in the coming days. I will keep you updated how it goes.

vj4git commented:
I added a new dtsi and also added that with other relevant files looking at  @jonathan-sturges latest commit. The initramfs build was successful, but for some reason when I try to upgrade via OEM firmware GUI (step #1 instructions above) I am getting "Upgrade Firmware Error!", after 7/8 seconds of the bar. No further error or log.

Does it matter how name the model? I called it  "Amped Wireless ALLY12B" (for my AR1200L) since their manual refers that way.

jonathan-sturges commented:
It definitely did not like the new image.  The device name shouldn't matter.  The R1900K/00X19K require a CSYS header, which gets added by the KERNEL_INITRAMFS macro in the make file.  Assuming you started with that, comment that out and build again.  Maybe the AR1200L doesn't require CSYS?

vj4git commented:
I removed the CSYS and rebuilt. The target initramfs looked good, but it still didn't make any difference when uploading to the router. It reports Upgrade Firmware Error! Any more ideas? Is it possible to put the router under TFTP recovery mode?

jonathan-sturges commented:
AFAIK you can do TFTP recovery on the router the same as the 00X19K extender.

vj4git commented:
Ya, tried that before. Strangely, it is not reachable with ping after that either as 192.168.3.1 or 192.168.1.1 (tftp times out) - of course, my laptop is appropriate network segment (.3 or .1)

jonathan-sturges commented:
Try 192.168.1.6, the 00X19K uses that when in TFTP recovery mode.

vj4git commented:
LOL. Spoke too early. Yes, you're right - 192.168.1.6 worked and I was able to tftp successfully. I could see that it rebooted. However, now I don't know what address it has. 192.168.1.1 or 1.6 or 3.1 don't connect to.

jonathan-sturges commented:
It should be .1.1 after rebooting into OpenWRT.

jonathan-sturges commented:
I forgot to point out, use the CSYS-wrapped image for TFTP recovery.  If it's like the ALLY19 devices, it'll likely be rejected without CSYS during recovery too.

vj4git commented:
Yes, I observed that. After TFTP recovery I was not able to ping 192.168.1.1. Then I did the TFTP recovery with my earlier CSYS flag image and then viola - I could ping the box with 192.168.1.1.

Now, my other issue - no openwrt site pulling up. What do you think the culprit could be?

vj4git commented:
I am able to ssh to the box. My openwrt debug skills are limited - I don't find anything on /var/log

jonathan-sturges commented:
Cool that you got in.  You probably have to go back and explicitly enable LuCI in menuconfig in order to get it into the initramfs image.  While you're in with ssh, save off the output of 'dmesg' though.

vj4git commented:
Thanks @jonathan-sturges - Enabling LuCI first time somehow not worked with the initramfs and had trouble getting it on the box. After several attempts, I cleaned up the whole openwrt image, rebuilt from scratch and TFTP (with CSYS header). Finally, I am good and running on the openwrt GUI for my AR1200L. My current Linksys WRT1900acs box has several openwrt configs, multiple SSID, VLANs, firewall zones etc and I will have to repeat all those and check it out. If good, I will proceed with the "Extender".

Question - I am a longterm openwrt user, but this is my first build. Do I always need to rebuild my own firmware every time moving forward?

jonathan-sturges commented:
Cool!  I would recommend backing up the OEM firmware so you can go back if need be... see my commit note.  Also, while running the initramfs image, you have no overlay so you cannot save your config or any add'l packages.  The intent is to use the initramfs image long enough to sysupgrade to a proper image that has an overlay filesystem.  I would hold off doing this until you're sure everything works.  If everything's 100%, submit your .dts and makefile changes to make your device official, then you won't need to build from scratch.

vj4git commented:
Sorry - question on wlan. My 2.4G works, but not the 5G. I am trying to understand the dts. Does your R-1900K have both internal and external amplifiers? This is what the AR1200L techspecs manual says:

Amplifier: 2 x 2.4GHz External Amplifiers
 2 x 5.0GHz External Amplifiers
 4 x Internal Low Noise Amplifiers

I did a hexdump of /dev/mtd2 (factory) and the address 0x00004 has the actual mac address of my 5G. However, I see only 2.4G is active after boot.

Should I make any change to the &pcie1 section of the dtsi?

jonathan-sturges commented:
The idea is to have a dtsi for common components of the router & extender and then separate dts for each device.  Make new dtsi/dts files for your devices if they're different than the 19's.

jonathan-sturges commented:
My ALLYs actually have their radio MACs at 0x4 and 0x8004 too, but every other mt7621 device uses 0x0 and 0x8000 in their DTS.  In practice, it doesn't seem to matter, the radios pick up their correct MACs; it's the ethernets that need help.  BTW all MACs are in ASCII in the hwconfig partition.

adschm commented:
0x0/0x8000 is the _calibration data_, and 0x4/0x8004 are the _MAC adresses_ with an offset of +4 inside the calibration data (for this architecture). For ethernet, you then typically have additional addresses at 0x28/0x2e and 0xe000/0xe006 ...

vj4git commented:
Ok  @jonathan-sturges @adschm - that makes sense. There are two issues I am dealing with
i) I am stuck with the initramfs.bin uploaded via TFTP and when I try the sysupgrade with squashfs via GUI or CLI it seems to be working, but after reboot I only see the old build version. For example, I change the root password and after sysupgrade it is lost. When I do df -h, I only see /tmp and /dev on tmpfs and no other.

- my dmesg reports:
"UBI error: no valid UBI magic found inside mtd5"
I believe this could be my issue for not able to write the squashfs. 

ii) dmesg reports mt-7621 pci 1e140000.pcie: Error applying setting, reverse things back. Is this causing the 5G issue

Appreciate if you have any ideas.

jonathan-sturges commented:
@vj4git The UBI error is expected when on the initramfs, but should go away after sysupgrade.  The 1st sysupgrade will fail unless you tell it not to save the configuration.  And for the DT error you mention, I've always seen that but it doesn't have any effect I can find.

vj4git commented:
Here is the dmesg on 7615:

> oot@OpenWrt:~# dmesg | grep 7615
[    6.560653] pci 0000:01:00.0: [14c3:7615] type 00 class 0x000280
[   13.752552] mt7615e 0000:01:00.0: enabling device (0000 -> 0002)
[   13.793188] mt7615e 0000:01:00.0: HW/SW Version: 0x8a108a10, Build Time: 20180518100604a
[   13.836855] mt7615e 0000:01:00.0: N9 Firmware Version: _reserved_, Build Time: 20200814163649
[   13.875011] mt7615e 0000:01:00.0: CR4 Firmware Version: _reserved_, Build Time: 20190121161307
[   21.545540] Modules linked in: pppoe ppp_async iptable_nat xt_state xt_nat xt_conntrack xt_REDIRECT xt_MASQUERADE xt_FLOWOFFLOAD pppox ppp_generic nf_nat nf_flow_table_hw nf_flow_table nf_conntrack_rtcache nf_conntrack mt7615e mt7615_common mt76 mac80211 ipt_REJECT cfg80211 xt_time xt_tcpudp xt_multiport xt_mark xt_mac xt_limit xt_comment xt_TCPMSS xt_LOG slhc nf_reject_ipv4 nf_log_ipv4 nf_defrag_ipv6 nf_defrag_ipv4 iptable_mangle iptable_filter ip_tables crc_ccitt compat ledtrig_heartbeat nf_log_ipv6 nf_log_common ip6table_mangle ip6table_filter ip6_tables ip6t_REJECT x_tables nf_reject_ipv6 leds_gpio gpio_button_hotplug
[   21.839777] [<865031e4>] mt7615_register_ext_phy+0x274/0x2ac [mt7615_common]

Radio itself seems to be ok, because in the GUI I can activate for 2.4g (only N). I am not seeing 5g AC and wonder what do I need to make it active.

RE: sysupgrade - the issue is same if I try 

> sysupgrade -n -v /tmp/squashfs-sysupgrade.bin

It says upgrading and then reboots, but it comes back alive it is my original tftp uploaded initramfs. Just for testing I also did the direct 

> mtd -r write /tmp/squashfs-sysupgrade.bin firmware

same behavior.

vj4git commented:
I checked within make menuconfig both kmod-mt7615-firmware and kmod-mt7615e are selected, so driver should not be an issue. Strangely, the MAC address shown under openwrt wireless radio0 is really 5G's MAC as per router sticker but the GUI shows it as 2G. Per the label the 2G MAC is to be increment by 1 to this.

jonathan-sturges commented:
Yeah not a driver problem.  The `pcie0` & `pcie1` DTS stanzas look OK?  If you can get an `lspci` output that would help.

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally-00x19k.dts:
```diff
@@ -0,0 +1,23 @@
+#include "mt7621_ampedwireless_ally.dtsi"
+
+/ {
+	compatible = "ampedwireless,ally-00x19k", "mediatek,mt7621-soc";
```
I wonder whether we should just use "amped" for the vendor part here ...

jonathan-sturges replied:
I thought about that but ultimately decided this was unambiguous.

jonathan-sturges replied:
@adschm FWIW, elsewhere in ramips we have vendors "alfa-network", "alphanetworks", "mofinetwork", "rakwireless" and "arcwireless".

adschm commented on package/boot/uboot-envtools/files/ramips:
```diff
@@ -25,6 +25,10 @@ allnet,all0256n-8m|\
 	;;
+ampedwireless,ally-00x19k|\
+ampedwireless,ally-r1900k)
+	ubootenv_add_uci_config "/dev/mtd1" "0x0" "0x1000" "0x20000" "4"
```
What's the "4"?

jonathan-sturges replied:
The uboot-env partition on these devices is massive (512KiB) relative to the actual environment size.  There are 4 0x20000 sectors in the partition.

jonathan-sturges replied:
The u-boot-env on this device is huge:  512KiB.  Those options to ubootenv_add_uci_config() are the sector size & num sectors.

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally-00x19k.dts:
```diff
@@ -0,0 +1,23 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
```
dts-v1 can be dropped now, it has been moved to mt7621.dtsi

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally-00x19k.dts:
```diff
@@ -0,0 +1,23 @@
+
+&xhci {
+	status = "disabled";
+};
```
Add newline her.

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally-r1900k.dts:
```diff
@@ -0,0 +1,34 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
```
dts-v1 can be dropped now, it has been moved to mt7621.dtsi

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally-r1900k.dts:
```diff
@@ -0,0 +1,34 @@
+			label = "lan3";
+		};
+	};
+};
```
Add newline here.

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally.dtsi:
```diff
@@ -0,0 +1,184 @@
+			read-only;
+		};
+	};
+};
```
Missing newline ...

jonathan-sturges replied:
ok

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -140,6 +140,33 @@ define Device/alfa-network_quad-e4g
+  KERNEL_INITRAMFS := $(KERNEL_DTB) | custom-initramfs-uimage flashable-initramfs |\
+     edimax-header -s CSYS -m RN68 -f 0x001c0000 -S 0x01100000
+  IMAGE/sysupgrade.bin := sysupgrade-tar | append-metadata
+  DEVICE_VENDOR := Amped Wireless
```
Please order the variables more like the other devices.

jonathan-sturges replied:
Honestly I don't see a lot of consistency throughout the mt7621.mk, but I gave it a shot.

adschm replied:
My main point is that DEVICE_VENDOR should be at the top, not at the bottom....

jonathan-sturges replied:
ok thanks, I think this should be better

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -140,6 +140,33 @@ define Device/alfa-network_quad-e4g
+  KERNEL_SIZE := 4096k
+  IMAGE_SIZE := 32768k
+  KERNEL_INITRAMFS := $(KERNEL_DTB) | custom-initramfs-uimage flashable-initramfs |\
+     edimax-header -s CSYS -m RN68 -f 0x001c0000 -S 0x01100000
```
Hanging indent with one tab.

vj4git commented:
@jonathan-sturges - below is the lspci output:

> 00:00.0 PCI bridge: Device 0e8d:0801 (rev 01)
> 01:00.0 Unclassified device [0002]: MEDIATEK Corp. Device 7615

jonathan-sturges commented:
@vj4git I'm not sure why the 2nd radio doesn't show up, but it feels like something's likely wrong with the PCIe setup in the DTS.

vj4git commented:
I looked at various combinations in the dts and the only thing that is close to the hardware is below (similar to how elecom - mt7621_elecom_wrc-1167ghbk2-s.dts). What I have is a "single" radio, but supporting dual Bands (11g and 11ac).
```
&pcie {
	status = "okay";
};
&pcie0 {
	wifi@0,0 {
		compatible = "mediatek,mt76";
		reg = <0x0000 0 0 0 0>;
		mediatek,mtd-eeprom = <&factory 0x0>;
	};
};
```

In the elecom's case there are also below two but adding that produces a dmesg Invalid MAC Address error:
```
 		mtd-mac-address = <&factory 0xe000>;
 		mtd-mac-address-increment = <1>;
```

By changing &pcie0 as above (no &pcie1)  I am able to see both the bands in "iw list" with a single radio0. Any other configuration only reports a single band.

Within /etc/config/wireless I can setup the hwmode 11a  and I am able to connect to the 5g.

I have some devices that can only see 2.4g, but I would like to keep the 5g active whichever devices can. Appreciate if any ideas?

jonathan-sturges commented:
Glad you figured it out!  That's very similar to what I had; did you try just removing the ieee80211-freq-limit from the original dtsi?

vj4git commented:
Yes - your original configuration without the ieee80211-freq-limit lines work similarly and iw list shows two bands. So, I will keep the common dtsi without the ieee80211-freq-limit lines.

However, my problem is still there with all the configuration changes so far - this radio can be seen by the laptops/ios devices (that technically supports 5g), but the older devices (remotes, old laptops etc) don't even see this AP. I added hwmode to "11ag", but it doesn't seem to have any effect on older devices. I remember, with the stock firmware the older devices were able to see and join the AP. If I can get that working with openwrt it would be nice, rather than run another router for older devices.

jonathan-sturges commented:
One of the causes I've seen for that problem is the channel width; the "a" devices can't use any channel wider than 20MHz.  There could be other reasons, like security... try turning off security temporarily, or at least dumming it down from WPA2 to WPA to WEP for testing.

vj4git commented:
Tried that - even Open AP didn't show up on the older laptop

jonathan-sturges commented:
Beyond the above, for 2.4Ghz, make sure `Allow legacy 802.11b rates` is checked, since "b" clients expect beacons transmitted at 1Mbps.  After that, I'd check with the forums, someone else may have more ideas.

vj4git commented:
Allow legacy 802.11b rates are already checked in LuCI. I wish openwrt had some way of separating "Bands", rather than only as hardware radio. I will check on the forum.

jonathan-sturges commented:
Thinking about this more, you probably should setup a separate network for legacy devices anyway; pretty much anything prior to "n" will slow your network down and/or require reduced security.

vj4git commented:
Sure, my current Linksys WRT1900acv2 has two radios - so, I have the frequencies separated and only the recent gadgets on the 5g SSID.  I was thinking if there was a way to do similar separation leveraging "bands" concept.

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally.dtsi:
```diff
@@ -0,0 +1,184 @@
+		compatible = "gpio-leds";
+
+		led_status_green: status_green {
+			label = "ally:green:status";
```
Drop the "ally:" prefix for all LEDs, so just the color:function part remains.

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -140,6 +140,33 @@ define Device/alfa-network_quad-e4g
+  UBINIZE_OPTS := -E 5
+  FILESYSTEMS := squashfs ubifs
+  KERNEL_INITRAMFS := $(KERNEL_DTB) | custom-initramfs-uimage flashable-initramfs |\
+  edimax-header -s CSYS -m RN68 -f 0x001c0000 -S 0x01100000
```
Hanging indent should be a single tab only.

jonathan-sturges commented:
@adschm  It looks like there's a fair number of new Ramips devices being contributed.  Is the team intending for them to be included in OpenWRT 20.x?

adschm commented:
> @adschm It looks like there's a fair number of new Ramips devices being contributed. Is the team intending for them to be included in OpenWRT 20.x?

20.xx will contain whatever master contains when it is branched. Which device support PRs will be merged until then essentially is a matter of chance as always, depending on which PRs developers have time for and interest in. That's the way it works on a free-time project, and unfortunately I cannot promise you anything more substantial.

jonathan-sturges commented:
Updated mt7621.mk to remove use of OBE "custom-initramfs-uimage" function.

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally.dtsi:
```diff
@@ -0,0 +1,184 @@
+
+		partition@140000 {
+			label = "firmware";
+			reg = <0x140000 0x2000000>;	    /* 32MB */
```
Is the nesting necessary here, or could you just remove the firmware partition and have kernel and ubi at the same level as the other partitions?

jonathan-sturges replied:
Kernel & UBI partitions can be at the same level as the others.  I'd seen this style on other devices and just liked the readability.

jonathan-sturges replied:
Made the requested edit.

jonathan-sturges commented:
@adschm I remembered one need for the nested-partition layout in DTS:  fully imaging the OEM firmware.  In order to do so, users need a way to refer to that entire 32MB flash segment.  Is there an approach I can take that's more palatable than nesting my OpenWRT partitions?

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally-00x19k.dts:
```diff
@@ -0,0 +1,22 @@
+/ {
+	compatible = "ampedwireless,ally-00x19k", "mediatek,mt7621-soc";
+	model = "Amped Wireless ALLY-00X19K";
+
```
drop unneeded empty line.

jonathan-sturges replied:
done.

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally-r1900k.dts:
```diff
@@ -0,0 +1,33 @@
+/ {
+	compatible = "ampedwireless,ally-r1900k", "mediatek,mt7621-soc";
+	model = "Amped Wireless ALLY-R1900K";
+
```
drop unneeded empty line.

jonathan-sturges replied:
done.

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally.dtsi:
```diff
@@ -0,0 +1,179 @@
+};
+
+&pcie0 {
+	wifi0: wifi@0,0 {
```
Do we need the "wifi0" and "wifi1" DT labels? If not, please drop them.

jonathan-sturges replied:
I removed them.

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -114,6 +114,33 @@ define Device/alfa-network_quad-e4g
 endef
 TARGET_DEVICES += alfa-network_quad-e4g
 
+define Device/ampedwireless_ally
```
Consider naming this Device/ampedwireless_ally_common so it's obvious it's not a device itself.

jonathan-sturges replied:
good idea; done.

adschm commented on target/linux/ramips/mt7621/base-files/etc/board.d/02_network:
```diff
@@ -34,13 +21,41 @@ ramips_setup_interfaces()
+	xiaomi,mi-router-3g-v2|\
+	xiaomi,mi-router-4a-gigabit)
+		ucidef_set_interfaces_lan_wan "lan1 lan2" "wan"
+		;;
```
This is added twice?

jonathan-sturges replied:
whoops, fixed.

jonathan-sturges commented:
@adschm can you verify the changes when you have a minute?

jonathan-sturges commented:
@adschm can you confirm the changes when you have a minute?  TIA

adschm commented:
> jonathan-sturges

Your Signed-off-by might need some formatting ...

adschm commented on target/linux/ramips/mt7621/base-files/etc/board.d/02_network:
```diff
@@ -33,13 +20,27 @@ ramips_setup_interfaces()
+	mikrotik,routerboard-m33g|\
+	xiaomi,mi-router-3g|\
+	xiaomi,mi-router-3g-v2|\
+	xiaomi,mi-router-4a-gigabit)
```
One xiaomi is missing here...

jonathan-sturges replied:
Thanks for catching that, fixed.

jonathan-sturges commented:
> > jonathan-sturges
> 
> Your Signed-off-by might need some formatting ...

I thought you might be referencing line lengths.  I think longer line lengths are now allowed than in the past, so I reformatted the note accordingly.  If that's not what you had in mind, please let me know.

adschm commented:
> > > jonathan-sturges
> > 
> > 
> > Your Signed-off-by might need some formatting ...
> 
> I thought you might be referencing line lengths. I think longer line lengths are now allowed than in the past, so I reformatted the note accordingly. If that's not what you had in mind, please let me know.

I mean that your name is probably not "jonathan-sturges", but "Jonathan Sturges"? Is it really that hard?

jonathan-sturges commented:
> I mean that your name is probably not "jonathan-sturges", but "Jonathan Sturges"? Is it really that hard?

Not hard at all, and fixed.

adschm commented:
Hmm, I already picked this, but then I found that it's not setting ethernet MAC addresses at all anymore.

Apart from that, can you comment why this would be the only device in entire ramips that needs `FILESYSTEMS := squashfs ubifs` ?

adschm commented on target/linux/ramips/mt7621/base-files/etc/board.d/02_network:
```diff
@@ -34,13 +20,43 @@ ramips_setup_interfaces()
+	xiaomi,mi-router-4|\
+	xiaomi,mi-router-4a-gigabit)
+		ucidef_set_interfaces_lan_wan "lan1 lan2" "wan"
+		;;
```
And here, the blocks are added twice.... I will drop it again.

jonathan-sturges replied:
Fixed again.  Sigh.

jonathan-sturges commented:
> Apart from that, can you comment why this would be the only device in entire ramips that needs `FILESYSTEMS := squashfs ubifs` ?
It does not need that... now removed.

adschm commented:
Still no ethernet MAC addresses set. Do I really have to write everything twice?

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally.dtsi:
```diff
@@ -0,0 +1,179 @@
+	keys {
+		compatible = "gpio-keys";
+
+		led {
```
Please replace this by "led_switch"

jonathan-sturges replied:
ok

adschm commented on target/linux/ramips/dts/mt7621_ampedwireless_ally.dtsi:
```diff
@@ -0,0 +1,179 @@
+
+		partition@140000 {
+			label = "kernel";
+			reg = <0x140000 0x400000>;	/* 4MB */
```
remove the "size comments" here and below ...

jonathan-sturges replied:
ok

adschm commented on target/linux/ramips/image/mt7621.mk:
```diff
@@ -122,6 +122,32 @@ define Device/alfa-network_quad-e4g
 endef
 TARGET_DEVICES += alfa-network_quad-e4g
 
+define Device/ampedwireless_ally_common
```
Add $(Device/dsa-migration) here.

jonathan-sturges replied:
added.

adschm reviewed (commented):
Since this looks like it's gonna take another while, please also do these changes which I did myself earlier ...

jonathan-sturges committed:
ramips: add support for Amped Wireless ALLY router and extender

Amped Wireless ALLY is a whole-home WiFi kit, with a router (model
ALLY-R1900K) and an Extender (model ALLY-00X19K).  Both are devices are
11ac and based on MediaTek MT7621AT and MT7615N chips.  The units are
nearly identical, except the Extender lacks a USB port and has a single
Ethernet port.

Specification:
- SoC: MediaTek MT7621AT (2C/4T) @ 880MHz
- RAM: 128MB DDR3 (Nanya NT5CC64M16GP-DI)
- FLASH: 128MB NAND (Winbond W29N01GVSIAA)
- WiFi: 2.4/5 GHz 4T4R
  - 2.4GHz MediaTek MT7615N bgn
  - 5GHz MediaTek MT7615N nac
- Switch: SoC integrated Gigabit Switch
- USB: 1x USB3 (Router only)
- BTN: Reset, WPS
- LED: single RGB
- UART:  through-hole on PCB.
   J1: pin1 (square pad, towards rear)=3.3V, pin2=RX,
   pin3=GND, pin4=TX.  Settings: 57600/8N1.

Note regarding dual system partitions
-------------------------------------

The vendor firmware and boot loader use a dual partition scheme.  The boot
partition is decided by the bootImage U-boot environment variable: 0 for
the 1st partition, 1 for the 2nd.

OpenWrt does not support this scheme and will always use the first OS
partition.  It will set bootImage to 0 during installation, making sure
the first partition is selected by the boot loader.

Also, because we can't be sure which partition is active to begin with, a
2-step flash process is used.  We first flash an initramfs image, then
follow with a regular sysupgrade.

Installation:

Router (ALLY-R1900K)
1) Install the flashable initramfs image via the OEM web-interface.
  (Alternatively, you can use the TFTP recovery method below.)
  You can use WiFi or Ethernet.
  The direct URL is:  http://192.168.3.1/07_06_00_firmware.html
  a. No login is needed, and you'll be in their setup wizard.
  b. You might get a warning about not being connected to the Internet.
  c. Towards the bottom of the page will be a section entitled "Or
  Manually Upgrade Firmware from a File:" where you can manually choose
  and upload a firmware file.
  d: Click "Choose File", select the OpenWRT "initramfs" image and click
  "Upload."
2) The Router will flash the OpenWrt initramfs image and reboot.  After
  booting, LuCI will be available on 192.168.1.1.
3) Log into LuCI as root; there is no password.
4) Optional (but recommended) is to backup the OEM firmware before
  continuing; see process below.
5) Complete the Installation by flashing a full OpenWRT image.  Note:
  you may use the sysupgrade command line tool in lieu of the UI if
  you prefer.
  a.  Choose System -> Backup/Flash Firmware.
  b.  Click "Flash Image..." under "Flash new firmware image"
  c.  Click "Browse..." and then select the sysupgrade file.
  d.  Click Upload to upload the sysupgrade file.
  e.  Important:  uncheck "Keep settings and retain the current
      configuration" for this initial installation.
  f.  Click "Continue" to flash the firmware.
  g.  The device will reboot and OpenWRT is installed.

Extender (ALLY-00X19K)
1) This device requires a TFTP recovery procedure to do an initial load
  of OpenWRT.  Start by configuring a computer as a TFTP client:
  a. Install a TFTP client (server not necessary)
  b. Configure an Ethernet interface to 192.168.1.x/24; don't use .1 or .6
  c. Connect the Ethernet to the sole Ethernet port on the X19K.
2) Put the ALLY Extender in TFTP recovery mode.
  a. Do this by pressing and holding the reset button on the bottom while
  connecting the power.
  b. As soon as the LED lights up green (roughly 2-3 seconds), release
  the button.
3) Start the TFTP transfer of the Initramfs image from your setup machine.
For example, from Linux:
tftp -v -m binary 192.168.1.6 69 -c put initramfs.bin
4) The Extender will flash the OpenWrt initramfs image and reboot.  After
booting, LuCI will be available on 192.168.1.1.
5) Log into LuCI as root; there is no password.
6) Optional (but recommended) is to backup the OEM firmware before
  continuing; see process below.
7) Complete the Installation by flashing a full OpenWRT image.  Note: you
may use the sysupgrade command line tool in lieu of the UI if you prefer.
  a.  Choose System -> Backup/Flash Firmware.
  b.  Click "Flash Image..." under "Flash new firmware image"
  c.  Click "Browse..." and then select the sysupgrade file.
  d.  Click Upload to upload the sysupgrade file.
  e.  Important:  uncheck "Keep settings and retain the current
      configuration" for this initial installation.
  f.  Click "Continue" to flash the firmware.
  g.  The device will reboot and OpenWRT is installed.

Backup the OEM Firmware:
-----------------------

There isn't any downloadable firmware for the ALLY devices on the Amped
Wireless web site. Reverting back to the OEM firmware is not possible
unless we have a backup of the original OEM firmware.

The OEM firmware may be stored on either /dev/mtd3 ("firmware") or
/dev/mtd6 ("oem").  We can't be sure which was overwritten with the
initramfs image, so backup both partitions to be safe.

  1) Once logged into LuCI, navigate to System -> Backup/Flash Firmware.
  2) Under "Save mtdblock contents," first select "firmware" and click
  "Save mtdblock" to download the image.
  3) Repeat the process, but select "oem" from the pull-down menu.

Revert to the OEM Firmware:
--------------------------
* U-boot TFTP:
  Follow the TFTP recovery steps for the Extender, and use the
  backup image.

* OpenWrt "Flash Firmware" interface:
  Upload the backup image and select "Force update"
  before continuing.

Signed-off-by: Jonathan Sturges <jsturges@redhat.com>

jonathan-sturges commented:
> Still no ethernet MAC addresses set. Do I really have to write everything twice?
> (Or am I overlooking it somewhere?)

WLAN MACs work fine without any code, but you are correct in that the ethernet MACs got accidentally dropped from ramips_setup_macs()... and now fixed.

adschm commented:
Okay, I'll take it.
